File: mzcd3-code/code/Spotlight/PPRDataController.m
/***
 * Excerpted from "Core Data in Objective-C, Third Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/mzcd3 for more book information.
***/
//
//  PPRDataController.m
//  PPRecipes
//
//  Created by Marcus S. Zarra on 10/15/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

#import "PPRDataController.h"

@interface PPRDataController()

@property (nonatomic, copy) void (^initBlock)(void);

- (void)initializeCoreDataStack;

@end

@implementation PPRDataController

- (instancetype)init
{
  return [self initWithCompletion:NULL];
}

- (instancetype)initWithCompletion:(void(^)(void))block
{
  if (!(self = [super init])) return nil;

  [self setInitBlock:block];

  [self initializeCoreDataStack];

  return self;
}

- (void)saveContext
{
  NSManagedObjectContext *moc = [self managedObjectContext];

  if (!moc) return;
  if (![moc hasChanges]) return;

  NSError *error = nil;
  if (![moc save:&error]) {
    NSLog(@"Error saving MOC: %@\n%@", [error localizedDescription],
          [error userInfo]);
    abort();
  }
}

- (void)populateTypeEntities
{
  NSManagedObjectContext *moc = nil;
  NSManagedObjectContextConcurrencyType type = NSPrivateQueueConcurrencyType;
  moc = [[NSManagedObjectContext alloc] initWithConcurrencyType:type];
  [moc setParentContext:[self managedObjectContext]];
  [moc performBlockAndWait:^{
    NSFetchRequest *request = nil;
    request = [NSFetchRequest fetchRequestWithEntityName:@"Type"];

    NSError *error = nil;
    NSArray *result = [moc executeFetchRequest:request
                                         error:&error];
    NSAssert(result != nil, [error localizedDescription]);
    if ([result count]) return;

    //The types table has not been populated
    NSArray *types;
    types = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"RecipeTypes"];

    for (NSString *type in types) {
      NSManagedObject *object = [NSEntityDescription
                                 insertNewObjectForEntityForName:@"Type"
                                 inManagedObjectContext:moc];
      [object setValue:type forKey:@"name"];
    }
    if (![moc save:&error]) {
      NSLog(@"Failed to populate type: %@\n%@",
            [error localizedDescription], [error userInfo]);
      abort();
    }
  }];
}

#pragma mark - Core Data stack

- (void)initializeCoreDataStack
{
  NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"DataModel"
                                            withExtension:@"momd"];
  NSAssert(modelURL != nil, @"Failed to find model URL");

  NSManagedObjectModel *mom = nil;
  mom = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];

  NSPersistentStoreCoordinator *psc = nil;
  psc = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mom];

  NSManagedObjectContext *moc = nil;
  NSUInteger type = NSMainQueueConcurrencyType;
  moc = [[NSManagedObjectContext alloc] initWithConcurrencyType:type];
  [moc setPersistentStoreCoordinator:psc];

  [self setManagedObjectContext:moc];

  dispatch_queue_t queue;
  queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
  dispatch_async(queue, ^{
    NSMutableDictionary *options = [[NSMutableDictionary alloc] init];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSMigratePersistentStoresAutomaticallyOption];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSInferMappingModelAutomaticallyOption];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *cloudURL = [fileManager URLForUbiquityContainerIdentifier:nil];
    if (cloudURL) {
      cloudURL = [cloudURL URLByAppendingPathComponent:@"PPRecipes"];
      [options setValue:[[NSBundle mainBundle] bundleIdentifier]
                 forKey:NSPersistentStoreUbiquitousContentNameKey];
      [options setValue:cloudURL
                 forKey:NSPersistentStoreUbiquitousContentURLKey];
    }

    NSURL *sURL = nil;
    sURL = [[fileManager URLsForDirectory:NSDocumentDirectory
                                inDomains:NSUserDomainMask] lastObject];
    sURL = [sURL URLByAppendingPathComponent:@"PPRecipes-iCloud.sqlite"];
    NSError *error = nil;
    NSPersistentStoreCoordinator *coordinator = nil;
    coordinator = [[self managedObjectContext] persistentStoreCoordinator];
    NSPersistentStore *store = nil;
    store = [coordinator addPersistentStoreWithType:NSSQLiteStoreType
                                      configuration:nil
                                                URL:sURL
                                            options:options
                                              error:&error];
    if (!store) {
      NSLog(@"Error adding persistent store to coordinator %@\n%@",
           [error localizedDescription], [error userInfo]);
      abort();
      //Present a user facing error
    }

    [self populateTypeEntities];

    if ([self initBlock]) {
      dispatch_sync(dispatch_get_main_queue(), ^{
        [self initBlock]();
      });
    }
  });
}

@end

File: mzswift-code/code/Shared/iOS/PPRecipes/PPRDataController.swift
//
//  PPRDataController.swift
//  PPRecipes
//
//  Created by Marcus S. Zarra on 12/7/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

import Foundation
import CoreData

let RECIPE_TYPES = "ppRecipeTypes"

class PPRDataController: NSObject {
  var mainContext: NSManagedObjectContext?
  var writerContext: NSManagedObjectContext?
  var persistenceInitialized = false
  var initializationComplete: ((Error?) -> Void)?

  init(completion: @escaping (Error?) -> Void) {
    super.init()
    initializationComplete = completion
    initializeCoreDataStack()
  }

  func initializeCoreDataStack() {
    guard let modelURL = Bundle.main.url(forResource: "PPRecipes",
      withExtension: "momd") else {
        fatalError("Failed to locate DataModel.momd in app bundle")
    }
    guard let mom = NSManagedObjectModel(contentsOf: modelURL) else {
      fatalError("Failed to initialize MOM")
    }
    let psc = NSPersistentStoreCoordinator(managedObjectModel: mom)

    let type = NSManagedObjectContextConcurrencyType.mainQueueConcurrencyType
    mainContext = NSManagedObjectContext(concurrencyType: type)
    mainContext?.persistentStoreCoordinator = psc

    let queue = DispatchQueue.global(qos: .background)
    queue.async {
      let fileManager = FileManager.default
      guard let documentsURL = fileManager.urls(for: .documentDirectory,
        in: .userDomainMask).first else {
          fatalError("Failed to resolve documents directory")
      }
      let storeURL = documentsURL.appendingPathComponent("PPRecipes.sqlite")

      do {
        let options = [NSMigratePersistentStoresAutomaticallyOption: true,
          NSInferMappingModelAutomaticallyOption: true]
        try psc.addPersistentStore(ofType: NSSQLiteStoreType,
          configurationName: nil, at: storeURL, options: options)
      } catch let error {
        assertionFailure("Failed to add persistent store: \(error)")
        self.initializationComplete?(error)
      }
      self.populateTypeEntities()
      self.persistenceInitialized = true
    }
  }

  fileprivate func populateTypeEntities() {
    let pMOC = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
    pMOC.parent = mainContext
    pMOC.performAndWait() {
      let fetch = NSFetchRequest<NSManagedObject>(entityName: "Type")

      do {
        let count = try pMOC.count(for: fetch)
        if count > 0 { return }
      } catch {
        fatalError("Failed to count receipe types: \(error)")
      }

      guard let types = Bundle.main.infoDictionary?[RECIPE_TYPES] as? [String] else {
        fatalError("Failed to find default RecipeTypes in Info.plist")
      }

      for type in types {
        let object = NSEntityDescription.insertNewObject(forEntityName: "Type", into: pMOC)
        object.setValue(type, forKey:"name")
      }

      do {
        try pMOC.save()
      } catch {
        fatalError("Failed to save private child moc: \(error)")
      }
    }
  }

  func saveContext() {
    guard let main = mainContext else {
      fatalError("save called before mainContext is initialized")
    }
    main.performAndWait({
      if !main.hasChanges { return }
      do {
        try main.save()
      } catch {
        fatalError("Failed to save mainContext: \(error)")
      }
    })
    guard let writer = writerContext else {
      return
    }
    writer.perform({
      if !writer.hasChanges { return }
      do {
        try writer.save()
      } catch {
        fatalError("Failed to save writerContext: \(error)")
      }
    })
  }
}

